File: Algorithms/725. Split Linked List in Parts.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn split_list_to_parts(head: Option<Box<ListNode>>, k: i32) -> Vec<Option<Box<ListNode>>> {
        let mut head = head;
        let mut count = 0;
        let mut cur = head.as_ref();
        while let Some(node) = cur {
            count += 1;
            cur = node.next.as_ref();
        }

        let (mut part, mut remainder) = (count / k, count % k);
        let mut res = vec![None; k as usize];

        for i in 0..(k as usize) {
            res[i] = head;
            let mut cur = res[i].as_mut();
            for j in 1..(part + if remainder > 0 {
            remainder -= 1;
            1
            } else {
            0
            }) {
                cur = cur.unwrap().next.as_mut();
                }
                if let Some(node) = cur {
                head = node.next.take();
                } else {
                break;
            }
        }
        res
    }
}

File: Algorithms/725. Split Linked List in Parts.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function splitListToParts(
  head: ListNode | null,
  k: number
): Array<ListNode | null> {
  let total = 0;
  let cur = head;
  while (cur) {
    total += 1;
    cur = cur.next;
  }
  const result: ListNode[] = [];
  const quo = Math.floor(total / k);
  let rem = total % k;
  const groups = new Array(k).fill(quo);
  for (let i = 0; i < rem; i++) {
    groups[i] += 1;
  }

  for (const group of groups) {
    if (group === 0) {
      result.push(null);
      continue;
    }
    let node = new ListNode(head.val);
    result.push(node);
    for (let i = 0; i < group - 1; i++) {
      head = head.next;
      node.next = new ListNode(head.val);
      node = node.next;
    }
    head = head.next;
  }
  return result;
}
